### PR TITLE
Detect and remove all duplicate pixels

### PR DESCRIPTION
**Pull request by fwyzard**
#### PR description:

Detect and remove all duplicate pixels, after unpacking each pixel module but before running the clustering.
Use shared memory for inter-thread communication and to speed up marking and detecting the duplicates.

#### PR validation:

Running the online pixel reconstruction and the full HLT menu on GPU over non-problematic events shows only a moderate slow down:

reconstruction | pixel tracking | full HLT
---|---|---
no duplicate removal                                                          |1566 ± 16 ev/s (--)    | 873 ± 4 ev/s (--)
duplicate removal with `atomicOR` (ce8a57bd815d631764ada169206d0eeb0e33b09e)  |1530 ± 17 ev/s (-2.3%) | 872 ± 4 ev/s (-0.2%)
duplicate removal with `atomicCAS` (25220129fa7dc0e5caac01f2800234e2ad956877) |1519 ± 14 ev/s (-3.0%) | 869 ± 2 ev/s (-0.4%)

#### If this PR will be backported please specify to which release cycle the backport is meant for:

To be backported to 12.4.x for data taking (#38947).

### CODE REVIEW

**Comment by fwyzard** [comment]
type bugfix

**Comment by fwyzard** [comment]
enable gpu

**Comment by fwyzard** [comment]
please test

**Comment by fwyzard** [comment]
urgent

**Comment by fwyzard** [comment]
@VinInn @AdrianoDee FYI

**Comment by fwyzard** [comment]
+heterogeneous

**Comment by civanch** [comment]
+1

File: RecoLocalTracker/SiPixelClusterizer/plugins/gpuClustering.h [hunk 1]
```diff
@@ -7,10 +7,54 @@
...
+  namespace pixelStatus {
+    // Use 0x00, 0x01, 0x03 so each can be OR'ed on top of the previous ones
+    enum Status : uint32_t { kEmpty = 0x00, kFound = 0x01, kDuplicate = 0x03 };
+
+    constexpr uint32_t bits = 2;
+    constexpr uint32_t mask = (0x01 << bits) - 1;
+    constexpr uint32_t valuesPerWord = sizeof(uint32_t) * 8 / bits;
+    constexpr uint32_t size = pixelSizeX * pixelSizeY / valuesPerWord;
+
+    __device__ static constexpr inline uint32_t getIndex(uint16_t x, uint16_t y) {
```
Comment (VinInn) [inline RecoLocalTracker/SiPixelClusterizer/plugins/gpuClustering.h:28]: why __device__ ?

File: RecoLocalTracker/SiPixelClusterizer/plugins/gpuClustering.h [hunk 2]
```diff
@@ -39,14 +83,19 @@ namespace gpuClustering {
...
                            uint16_t const* __restrict__ y,            //
                            uint32_t const* __restrict__ moduleStart,  // index of the first pixel of each module
                            uint32_t* __restrict__ nClustersInModule,  // output: number of clusters found in each module
                            uint32_t* __restrict__ moduleId,           // output: module id of each module
                            int32_t* __restrict__ clusterId,           // output: cluster id of each pixel
                            int numElements) {
+    // status is only used for Phase-1, but it cannot be declared conditionally only if isPhase2 is false;
+    // to minimize the impact on Phase-2 reconstruction it is declared with a very small size.
+    constexpr const uint32_t pixelStatusSize = isPhase2 ? 1 : pixelStatus::size;
+    __shared__ uint32_t status[pixelStatusSize];  // packed words array used to store the PixelStatus of each pixel
```
Comment (VinInn) [inline RecoLocalTracker/SiPixelClusterizer/plugins/gpuClustering.h:98]: if we have now this map, one may remove the Histo and the neighbor array and use that...
(it needs to be updated for Phase2).

File: RecoLocalTracker/SiPixelClusterizer/plugins/gpuClustering.h [hunk 3]
```diff
@@ -7,10 +7,54 @@
...
+
+    __device__ constexpr inline void promote(uint32_t* __restrict__ status, const uint16_t x, const uint16_t y) {
+      uint32_t index = getIndex(x, y);
+      uint32_t shift = getShift(x, y);
+      Status old = getStatus(status, x, y);
+      if (kDuplicate == old) {
+        // nothing to do
+        return;
+      }
+      Status target = (kEmpty == old) ? kFound : kDuplicate;
```
Comment (VinInn) [inline RecoLocalTracker/SiPixelClusterizer/plugins/gpuClustering.h:52]: ia this race free?

File: RecoLocalTracker/SiPixelClusterizer/plugins/gpuClustering.h (hunk 1 above)
Comment (VinInn) [reply RecoLocalTracker/SiPixelClusterizer/plugins/gpuClustering.h:28]: ```atomicOr``` needs to be added to cudaCompat

File: RecoLocalTracker/SiPixelClusterizer/plugins/gpuClustering.h (hunk 2 above)
Comment (fwyzard) [reply RecoLocalTracker/SiPixelClusterizer/plugins/gpuClustering.h:98]: Good point. I would rather not try to do it myself, though :-)

For Phase-2, unfortunately the modules are too large to have a map in shared memory :-(

File: RecoLocalTracker/SiPixelClusterizer/plugins/gpuClustering.h (hunk 3 above)
Comment (fwyzard) [reply RecoLocalTracker/SiPixelClusterizer/plugins/gpuClustering.h:52]: No :-(

We could have two threads setting the same pixel from `kEmpty` to `kFound` at the same time, and not detect that it is a duplicate.

File: RecoLocalTracker/SiPixelClusterizer/plugins/gpuClustering.h (hunk 1 above)
Comment (fwyzard) [reply RecoLocalTracker/SiPixelClusterizer/plugins/gpuClustering.h:28]: Unfortunately yes, I think you are right.

File: RecoLocalTracker/SiPixelClusterizer/plugins/gpuClustering.h (hunk 3 above)
Comment (clacaputo) [reply RecoLocalTracker/SiPixelClusterizer/plugins/gpuClustering.h:52]: > No :-(
> 
> We could have two threads setting the same pixel from `kEmpty` to `kFound` at the same time, and not detect that it is a duplicate.

hi @fwyzard, is it something that you are planning to fix in this PR?

Comment (fwyzard) [reply RecoLocalTracker/SiPixelClusterizer/plugins/gpuClustering.h:52]: Yes ...

Comment (clacaputo) [reply RecoLocalTracker/SiPixelClusterizer/plugins/gpuClustering.h:52]: Ok, thanks

File: RecoLocalTracker/SiPixelClusterizer/plugins/gpuClustering.h [hunk 4]
```diff
@@ -7,10 +7,54 @@
...
+      atomicOr(&status[index], static_cast<uint32_t>(target) << shift);
```
Comment (fwyzard) [inline RecoLocalTracker/SiPixelClusterizer/plugins/gpuClustering.h:53]:
```suggestion
+      uint32_t old_word = status[index];
+      uint32_t expected = old_word;
+      do {
+        expected = old_word;
+        Status old_status{(old_word >> shift) & mask};
+        if (kDuplicate == old_status) {
+          // nothing to do
+          return;
+        }
+        Status new_status = (kEmpty == old_status) ? kFound : kDuplicate;
+        uint32_t new_word = old_word | (static_cast<uint32_t>(new_status) << shift);
+        old_word = atomicCAS(&status[index], expected, new_word);
+      } while (expected != old_word);
```

I'm not testing this version, with an `atomicCAS` instead of an `atomicOR`.

Comment (fwyzard) [reply RecoLocalTracker/SiPixelClusterizer/plugins/gpuClustering.h:53]: @VinInn does this look good to you ?

**Comment by fwyzard** [comment]
please test

File: RecoLocalTracker/SiPixelClusterizer/plugins/gpuClustering.h (hunk 4 above)
Comment (VinInn) [reply RecoLocalTracker/SiPixelClusterizer/plugins/gpuClustering.h]: looks good (or at least very similar to what I would have coded myself)

**Comment by civanch** [comment]
+1

**Comment by fwyzard** [comment]
+heterogeneous

**Comment by fwyzard** [comment]
@clacaputo @jpata as this is something we would like to deploy online sooner rather than later, could you let me know if you have any concerns about, if you think we should involve directly the DPG, _etc._ ?

**Comment by clacaputo** [comment]
> @clacaputo @jpata as this is something we would like to deploy online sooner rather than later, could you let me know if you have any concerns about, if you think we should involve directly the DPG, _etc._ ?

No concerns from my side, just busy with other stuff. I'm going to sign it

**Comment by clacaputo** [comment]
+reconstruction

**Comment by fwyzard** [comment]
> No concerns from my side, just busy with other stuff.

Sure, no problem!

>  I'm going to sign it

Thanks.

**Comment by qliphy** [comment]
+1

**Comment by nothingface0** [comment]
@fwyzard Could you explain why this way is more efficient than the one proposed by @VinInn [here](https://github.com/cms-sw/cmssw/pull/37359)?

I can see that Vincenzo's way accesses the global memory `x` and `y` multiple times, and in the worst case those accesses are `numElements` apart. Does this mean that there will probably be many cache misses?

Is your way more efficient due to the global memory access being sequential (even though it's done twice?)

**Comment by fwyzard** [comment]
hi @nothingface0,
the main difference is that this approach is using _shared_ memory: https://github.com/cms-sw/cmssw/blob/c797f7f447cbc546e98aadb3d6819593de37b2ce/RecoLocalTracker/SiPixelClusterizer/plugins/gpuClustering.h#L105

Shared memory is local to the processor, like the L1 cache, so it is much faster than global memory.
The limitation is that there is only order of 48 - 64 kB available, so for example we cannot use the same approach for Phase-2, where the pixel modules are much larger.

**Comment by nothingface0** [comment]
I totally understood the use of shared memory in your approach, but I'm still a bit confused, since, for example, Vincenzo's approach does not need an extra array for storing occurrences. So it's not like the shared memory is used for existing data but for extra data.

To make my question clearer:

**Vincezo's approach:** Each thread does a pairwise comparison of the coordinates of the pixel it is assigned, with **every the coordinates of every other pixel**, i.e. the first thread compares pixel `0` with `1`, then `0` with `2`.. up to `0` with `numElements-1`, which are far apart in memory, meaning that this might lead to cache misses. Not to mention that this happens in parallel with every other thread trying to do its own comparisons. In general, this needs **`numElements` over 2** memory/cache accesses in total, whose pattern is not predictable.

**Your approach:** Use some extra (shared) memory to store occurrences of each pixel. To do that, you
* Sweep once over the `x` and `y` arrays sequentially (i.e. predictable for the GPU) to count occurrences of each pixel (`pixelStatus::promote(status, x[i], y[i]);`)
* Sweep a second time sequentially over the same arrays to check if they've been marked duplicates (`if (pixelStatus::isDuplicate(status, x[i], y[i]))`)

This means that you access the global memory at most **2 x `numElements`** times predictably, meaning better use of the available cache. 

TL;DR: Is this statement correct?:
Your approach is more efficient, not because of shared memory usage (since Vincenzo's method simply does not need an extra array), but because its global memory access patterns are more predictable and make better use of the GPU cache.  

Sorry for insisting, I'm just trying to understand [this statement](https://github.com/cms-sw/cmssw/pull/37359#issuecomment-1093011381):
> to be clear. this solution is NOT computational-sustainable.

**Comment by fwyzard** [comment]
IIUC, the approach used by Vincenzo compares every pixel with every other pixel, so the complexity of the algorithm grows with the square of the number of pixels in a module, _i.e._ is `O(N²)`.

The approach I used reads every pixel a fixed number of times, so the complexity gorws linerly with the number of pixels in a module, _i.e._ is `O(N)`.

The exact number of operations and memory accesses i not exactly N² or N, but those are the order of the leading terms in the two cases..

My approach does make use of an extra memory buffer, which normally would add a large cost; being able to keep that buffer in shared memory makes the cost acceptable. If you feel like it, it could be interesting to measure the impact of using a buffer in global memory instead (with a byte per pixel, without all the bitwise operations).

**Comment by nothingface0** [comment]
Right, there's the complexity of the approach, too. Thanks!

> measure the impact of using a buffer in global memory instead
I might ask you again on how to run profiling using patatrack :sweat_smile:
